### ISSUE TITLE
关于generator.decode 的并发问题

### ISSUE THREAD

**Issue by CallmeZhangChenchen** [InternLM/lmdeploy#3053]
这里有一个已经close的 issue  https://github.com/InternLM/lmdeploy/issues/2958
使用generator.decode 完成了 tts 工程里面 llm 的 decode 过程，现在单 batch已经完成了，结果和性能都很棒，
我简化一下过程
```
self.tm_model = tm.TurboMind.from_pretrained(pretrain_path, 
                    engine_config = TurbomindEngineConfig(max_batch_size=1,cache_max_entry_count=0.2))
self.generator = self.tm_model.create_instance()

for i in range(max_len):
     output = self.generator.decode(input_ids, 
                    steps=[lmdeploy_len], 
                    input_embeddings=input_embeddings,
                    input_embedding_ranges=input_embedding_ranges, 
                    sequence_start=(i == 0), 
                    sequence_end=False)[-1][-1]
     if top_ids == self.speech_token_size:
             break
```
现在我想使用一下max_batch_size=4，或者 Persistent Batch，来提升一下单卡下的并发量

所以想问一下，在上面代码逻辑下，怎样才能 利用上Persistent Batch的能力

我想了三点：
1.  self.generator = self.tm_model.create_instance()  是不是可以create多个
2. 是不是可以充 self.generator.decode这里下功夫
3. 因为这整个流程跟传统的大模型不太一样，所以到万不得已的时候才会考虑 lmdeploy serve api_server

**Comment by lvhan028** [comment]
在最新的0.7.0版本中，decode 接口拿掉了。
通过传入请求参数，可以获取生成 token 的 logits。
可参考文档：https://lmdeploy.readthedocs.io/en/latest/llm/pipeline.html#get-logits-for-generated-tokens

**Comment by CallmeZhangChenchen** [comment]
好的 感谢， 我将尝试一下

**Comment by CallmeZhangChenchen** [comment]
@lvhan028 Hi！ 打扰了
```
from lmdeploy import pipeline, GenerationConfig

pipe = pipeline('internlm/internlm2_5-7b-chat')

gen_config=GenerationConfig(output_logits='generation'
                            max_new_tokens=10)
response = pipe(['Hi, pls intro yourself', 'Shanghai is'],
                gen_config=gen_config)
logits = [x.logits for x in response]
```
想请教一下， pipe()这个函数的参数列表在哪看的呀，   
我现在是直接输入input_embeddings，没有经过 tokenizer，所以想看看有没有办法

**Comment by CallmeZhangChenchen** [comment]
好的 简单看了一下，其实我并不是必须要 获取生成 token 的 logits，只是现有工程的获取方式可能不太一样
关于 input_embeddings 我计划这样做一下
0. 修改网络中的 emb， 修改源码不用 tokenizer，直接传入 token，然后经过修改后的 emd的时候，可以解析出正确的 embeddings 
1. 或者可以  generator = inst.async_stream_infer(session_id, **kwargs)这个函数支持input_embeddings输入

2. 没有在代码中扒出来关于sampling_ids的代码，下面代码是我的top_id的计算方式，我可能需要适配一下，这段代码具体在哪个文件可以修改呀
```
            logp = output.log_softmax(dim=-1)
            top_ids = self.sampling_ids(logp.squeeze(dim=0), out_tokens, sampling, ignore_eos=True if i < min_len else False).item()
            if top_ids == self.speech_token_size:
                break
            if top_ids > self.speech_token_size:
                continue
            yield top_ids
            out_tokens.append(top_ids)
```
3. 其实我貌似只需要确定在0.7.0之前的版本，self.generator = self.tm_model.create_instance()   这里create多个instance，并且显存不增加，并发的时候互不影响，就简单很多了

**Comment by lvhan028** [comment]
如果使用 turbomind 引擎的话，sampling token ids 是在 csrc 里面的。
如果使用 pytorch 引擎，这个功能在 logits_process.py 里。

是想获取 logits，然后通过自定义的 logits process，生成 token 么？
